### PR TITLE
INFRASYS-5894: pseudo code on the rolling deploy

### PR DESCRIPTION
**Pull request by prabhakar1983**

### CODE REVIEW

File: License2Deploy/rolling_deploy.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+def update_ami_id_with_last_successfull():
+    """ AWS Call: As health check failed, update the ami_id with the last successfull one """
+    pass
+
+
+def rolling_deploy():
+    current_desired_ec2_count = return_autoscalinggroup_desired_size()
+    new_desired_ec2_count = calculate_new_desired_ec2_count(current_desired_ec2_count)
+    update_desired_ec2_count_in_autoscalinggroup(new_desired_ec2_count)
+    if (healthcheck_new_instances()):
```
Comment (mjgorman) [inline License2Deploy/rolling_deploy.py]: Need a function to check the status of the instance and wait for it to startup and become available for testing.

**Comment by mjgorman** [comment]
Needs to be done as a class. Should do things like connection in a the init for the class.

**Comment by mjgorman** [comment]
Should probably not rely on tags that don't exist, likely better to lambda filter get_all_groups() to find the correct autoscaling group object to use. 
```
import boto.ec2.autoscale as asg
conn = asg.connect_to_region('us-east-1')

conn.get_all_groups()
[AutoScaleGroup<awseb-e-gshhhkcvcx-stack-AWSEBAutoScalingGroup-6Y7UH1SZS8QM>, AutoScaleGroup<dnbi-backend-stg-dnbibillingASGstg-1RYWS3F6LT9PW>, AutoScaleGroup<dnbi-backend-stg-dnbicacheASGstg-1O3HO09P2AHYJ>, AutoScaleGroup<dnbi-backend-stg-dnbicirrusASGstg-1GRJQVHPS4COJ>, AutoScaleGroup<dnbi-backend-stg-dnbiemailASGstg-1T3DV362TS505>, AutoScaleGroup<dnbi-backend-stg-dnbigmsextenderASGstg-62N1XQ4MRHSH>, AutoScaleGroup<dnbi-backend-stg-dnbiorderfulfillmetASGstg-PRCOY9QT1MMQ>, AutoScaleGroup<php-autoscale-prd-asgphpautoscaleprd-LXII5VMCKRVQ>, AutoScaleGroup<php-autoscale-stg-asgphpautoscalestg-1QQWBX645V8N6>]

conn.get_all_groups()[1]
AutoScaleGroup<dnbi-backend-stg-dnbibillingASGstg-1RYWS3F6LT9PW>

conn.get_all_groups()[1].name
u'dnbi-backend-stg-dnbibillingASGstg-1RYWS3F6LT9PW'

conn.get_all_groups()[1].desired_capacity
2

>>> vars(conn.get_all_groups()[1])
{'default_cooldown': 300, 'instances': [Instance<id:i-fd0de128, state:InService, health:Healthy>, Instance<id:i-120ee2c7, state:InService, health:Healthy>], 'health_check_period': 0, 'created_time': u'2015-09-01T20:19:02.324Z', 'enabled_metrics': [], 'availability_zones': [u'us-east-1a'], 'member': u'\n      ', 'desired_capacity': 2, 'max_size': 10, 'placement_group': None, 'tags': [Tag(BUILD=0), Tag(ENV=stg), Tag(aws:cloudformation:logical-id=dnbibillingASGstg), Tag(aws:cloudformation:stack-id=arn:aws:cloudformation:us-east-1:186502235371:stack/dnbi-backend-stg/4dc05a60-50e6-11e5-9c61-50e24162947c), Tag(aws:cloudformation:stack-name=dnbi-backend-stg)], 'suspended_processes': [], 'autoscaling_group_arn': u'arn:aws:autoscaling:us-east-1:186502235371:autoScalingGroup:5c25a1fa-f7d6-40ca-bd6e-c835722c63af:autoScalingGroupName/dnbi-backend-stg-dnbibillingASGstg-1RYWS3F6LT9PW', 'min_size': 2, 'vpc_zone_identifier': u'subnet-492b283e', 'load_balancers': [u'dnbi-back-dnbibill-Z9JL3452TMBD'], 'launch_config_name': u'dnbi-backend-stg-dnbibillingLCstg-16T3K7X6XA5YW', 'name': u'dnbi-backend-stg-dnbibillingASGstg-1RYWS3F6LT9PW', 'termination_policies': [u'Default'], 'instance_id': None, 'connection': AutoScaleConnection:autoscaling.us-east-1.amazonaws.com, 'health_check_type': u'EC2'}

>>> len(conn.get_all_groups()[1].instances)
2
```

File: License2Deploy/rolling_deploy.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    return True
+
+
+def update_ami_id_with_last_successfull():
+    """ AWS Call: As health check failed, update the ami_id with the last successfull one """
+    pass
+
+def retrieve_auto_scaling_group(app_name, environment):
+    """ AWS Call: retrieve the Auto Scaling Group by Tag """
+    auto_scaling_groups = asg.connect_to_region("us-east-1").get_all_groups();
```
Comment (adamjkeller) [inline License2Deploy/rolling_deploy.py]: Connect be set in the constructor method and this should all be wrapped into a class.

As an example:

def __init__(self):
  self.conn = asg.connect_to_region('us-east'1)

then you just call self.conn.get_all_groups()

File: License2Deploy/rolling_deploy.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    return filter_auto_scaling_group(auto_scaling_groups, app_name, environment)
```
Comment (adamjkeller) [inline License2Deploy/rolling_deploy.py]: When defining the filter_auto_scaling_group function, you are looking for four values (auto_scaling_groups, app_name, project_name, environment), but are only passing in three here: (auto_scaling_groups, app_name, environment)

File: License2Deploy/rolling_deploy.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    auto_scaling_group.set_capacity(ec2_count)
+    """ AWS Call: update the desired_size of the AutoScaling Group of the Microserice"""
+    pass
+
+def healthcheck_new_instances():
+    """ AWS Call: Check whether the new instances are working good """
+    return True
+
+
+def update_ami_id_with_last_successfull():
```
Comment (adamjkeller) [inline License2Deploy/rolling_deploy.py]: I don't think this method is necessary. If the healtchecks fail, the job should return an exit code of something other than zero which will in turn trigger jenkins to run the job again with the last known ami id. The ami id should be passed in as a parameter to this script using sys.argv, argparse, etc.

**Comment by prabhakar1983** [comment]
@adamjkeller Corrected the items that you mentioned. thank you

**Pull request closed without merging**